### DISCUSSION TITLE
west flash for nucleo_u575zi_q is failing

### DISCUSSION THREAD

**Discussion by VineetaNarkhede-eaton** [zephyrproject-rtos/zephyr#44860] [General] [answered]
Hi,

I am following below document to build and flash hello_world sample code on nucleo_u575zi_q.
https://docs.zephyrproject.org/2.7.0/boards/arm/nucleo_u575zi_q/doc/nucleou575zi_q.html
Getting error while running west flash which uses stm32cubeprogrammer runner for flashing the code.

Error Output:
![image](https://user-images.githubusercontent.com/88422239/163183125-2eeaa04f-b31b-4900-a187-b918301f730a.png)

Commands used to build and flash:
```
west build -b nucleo_u575zi_q samples/hello_world
west flash
```

Able to run below command successfully but same is not working with west flash
`STM32_Programmer_CLI.exe --connect port=swd reset=HWrst --erase all`

![image](https://user-images.githubusercontent.com/88422239/163183411-ff7a6c7d-66af-411d-b63f-e57ebb606f8c.png)

OS: Windows
zephyr version: 2.7.99
STM32CubeProgrammer: 2.10.0

**Comment by FRASTM** [comment]
Did you update the Firmware of your nucleo board recently using the stm32CubeProgrammer ?That could help in any case 
![cubeProgrammer](https://user-images.githubusercontent.com/51417392/163211312-f7ed8f0d-cc70-41b3-aec1-5de52da1350d.JPG)

**Reply by VineetaNarkhede-eaton** [reply]
I tried Firmware upgrade, getting same error while running west flash after upgrade.

![image](https://user-images.githubusercontent.com/88422239/163327652-d94464c3-5a77-48dc-b513-f5f2b1ecef86.png)

![image](https://user-images.githubusercontent.com/88422239/163327749-9f69a1e4-b2e2-4d75-8b7c-275523fd9050.png)

![image](https://user-images.githubusercontent.com/88422239/163327999-5cd5501b-8cf2-4184-8929-530931a565f8.png)

**Comment by FRASTM** [comment]
Does your application run on a more recent zephyr Version : I see the _zephyr version: 2.7.99_
if possible try switching to `-- Zephyr version: 3.0.99`
I have Linux env but the $ west flash gives:
```
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner stm32cubeprogrammer
-- runners.stm32cubeprogrammer: mass erase requested
      -------------------------------------------------------------------
                        STM32CubeProgrammer v2.10.0                  
      -------------------------------------------------------------------

ST-LINK SN  : 002900274D46501120383832
ST-LINK FW  : V3J8M3
Board       : NUCLEO-U575ZI-Q
Voltage     : 3.29V
SWD freq    : 24000 KHz
Connect mode: Under Reset
Reset mode  : Hardware reset
Device ID   : 0x482
Revision ID : Rev X
Device name : STM32U575/STM32U585
Flash size  : 2 MBytes
Device type : MCU
Device CPU  : Cortex-M33
BL Version  : 0x92
Debug in Low Power mode enabled


Mass erase ... 

Mass erase successfully achieved
      -------------------------------------------------------------------
                        STM32CubeProgrammer v2.10.0                  
      -------------------------------------------------------------------

ST-LINK SN  : 002900274D46501120383832
ST-LINK FW  : V3J8M3
Board       : NUCLEO-U575ZI-Q
Voltage     : 3.29V
SWD freq    : 24000 KHz
Connect mode: Under Reset
Reset mode  : Hardware reset
Device ID   : 0x482
Revision ID : Rev X
Device name : STM32U575/STM32U585
Flash size  : 2 MBytes
Device type : MCU
Device CPU  : Cortex-M33
BL Version  : 0x92
Debug in Low Power mode enabled

      -------------------------------------------------------------------
        Choose flashing speed for Cortex M33 series.(default speed=Reliable)                  
      -------------------------------------------------------------------



Memory Programming ...
Opening and parsing file: zephyr.hex
  File          : zephyr.hex
  Size          : 15.25 KB 
  Address       : 0x08000000 


Erasing memory corresponding to segment 0:
Erasing internal memory sectors [0 1]
Download in Progress:
[==================================================] 100% 

File download complete
Time elapsed during download operation: 00:00:00.213

RUNNING Program ... 
  Address:      : 0x8000000
Application is running, Please Hold on...
Start operation achieved successfully
```

**Comment by FRASTM** [comment]
The same command is issued successfully by the west flash : `$ west -vvv flash`
```
-- runners.stm32cubeprogrammer: mass erase requested
runners.stm32cubeprogrammer: $CLI_PATH/STM32_Programmer_CLI --connect 'port=swd reset=HWrst' --erase all
```
and then 
`runners.stm32cubeprogrammer: $CLI_PATH/STM32_Programmer_CLI --connect 'port=swd reset=HWrst' --download ./zephyr/build/zephyr/zephyr.hex --start`

**Reply by VineetaNarkhede-eaton** [reply]
**On linux env west flash for nucleo_u575zi_q is running successfully at myside also, but I need to run it in Windows 10 env**
**I switched to recent zephyr version 3.0.99, getting same error,**
also tried `$ west -vvv flash`
```
(.venv) C:\Users\E0511444\zephyrsetup\latest\zephyr>west -vvv flash
args namespace: Namespace(command='flash', help=None, verbose=3, zephyr_base=None)
ZEPHYR_BASE=C:\Users\E0511444\zephyrsetup\latest\zephyr (origin: env)
config dir-fmt: build
build is a zephyr build directory
build dir: build
C:\Users\E0511444\zephyrsetup\latest\zephyr\build is a zephyr build directory
-- west flash: rebuilding
cmake version 3.22.2 is OK; minimum version is 3.13.1
Running CMake: 'C:\Program Files\CMake\bin\cmake.EXE' --build 'C:\Users\E0511444\zephyrsetup\latest\zephyr\build'
ninja: no work to do.
-- west flash: using runner stm32cubeprogrammer
runner_config: RunnerConfig(build_dir='C:\\Users\\E0511444\\zephyrsetup\\latest\\zephyr\\build', board_dir='C:/Users/E0511444/zephyrsetup/latest/zephyr/boards/arm/nucleo_u575zi_q', elf_file='C:\\Users\\E0511444\\zephyrsetup\\latest\\zephyr\\build\\zephyr\\zephyr.elf', hex_file='C:\\Users\\E0511444\\zephyrsetup\\latest\\zephyr\\build\\zephyr\\zephyr.hex', bin_file='C:\\Users\\E0511444\\zephyrsetup\\latest\\zephyr\\build\\zephyr\\zephyr.bin', gdb='C:/softwares/gnu_arm_embedded/bin/arm-none-eabi-gdb.exe', openocd='C:/softwares/OpenOCD-20211118-0.11.0/bin/openocd.exe', openocd_search=None)
-- runners.stm32cubeprogrammer: mass erase requested
runners.stm32cubeprogrammer: 'C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer\bin\STM32_Programmer_CLI.exe' --connect 'port=swd reset=HWrst' --erase all
FATAL ERROR: command exited with status 3221225477: 'C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer\bin\STM32_Programmer_CLI.exe' --connect 'port=swd reset=HWrst' --erase all
=== Traceback (enabled by -vvv):
Traceback (most recent call last):
  File "c:\users\e0511444\appdata\local\programs\python\python37\lib\site-packages\west\app\main.py", line 338, in run_command
    self.run_extension(args.command, argv)
  File "c:\users\e0511444\appdata\local\programs\python\python37\lib\site-packages\west\app\main.py", line 408, in run_extension
    command.run(args, unknown, self.topdir, manifest=self.manifest)
  File "c:\users\e0511444\appdata\local\programs\python\python37\lib\site-packages\west\commands.py", line 132, in run
    self.do_run(args, unknown)
  File "C:\Users\E0511444\zephyrsetup\latest\zephyr\scripts\west_commands\flash.py", line 29, in do_run
    do_run_common(self, my_args, runner_args)
  File "C:\Users\E0511444\zephyrsetup\latest\zephyr\scripts/west_commands\run_common.py", line 228, in do_run_common
    runner.run(command_name)
  File "C:\Users\E0511444\zephyrsetup\latest\zephyr\scripts/west_commands\runners\core.py", line 537, in run
    self.do_run(command, **kwargs)
  File "C:\Users\E0511444\zephyrsetup\latest\zephyr\scripts/west_commands\runners\stm32cubeprogrammer.py", line 173, in do_run
    self.flash(**kwargs)
  File "C:\Users\E0511444\zephyrsetup\latest\zephyr\scripts/west_commands\runners\stm32cubeprogrammer.py", line 195, in flash
    self.check_call(cmd + ["--erase", "all"])
  File "C:\Users\E0511444\zephyrsetup\latest\zephyr\scripts/west_commands\runners\core.py", line 640, in check_call
    subprocess.check_call(cmd, **kwargs)
  File "c:\users\e0511444\appdata\local\programs\python\python37\lib\subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['C:\\Program Files\\STMicroelectronics\\STM32Cube\\STM32CubeProgrammer\\bin\\STM32_Programmer_CLI.exe', '--connect', 'port=swd reset=HWrst', '--erase', 'all']' returned non-zero exit status 3221225477.
```

**Reply by FRASTM** [reply]
is your west up-to-date ? `$ west update `

**Reply by VineetaNarkhede-eaton** [reply]
Yes west is up-to-date, I did new setup today with latest zephyr version 3.0.99

**Reply by FRASTM** [reply]
It seems that the path to your cube STM32 programmer differs:
when you run manually the mass erase 
and when it is included in west flash
 `C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer\bin\STM32_Programmer_CLI.exe`

Could you check the STM32 programmer that the west flash is using ? 

What gives the 
C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer\bin\STM32_Programmer_CLI.exe --version

if not v2.10.0 Maybe you could try to re-install the STM32Cube Programmer V2.10.0 at the location where west flash is looking for (in your case = C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer\bin\STM32_Programmer_CLI.exe) 

Also Be sure to have the ST link of the target board not connected elsewhere  (by the windows CubeProgrammer application for example)

**Reply by VineetaNarkhede-eaton** [reply]
I have installed STM32CubeProgrammer at "C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer", same path is used by west flash runner. To confirm this I followed below steps:
1. Uninstall STM32CubeProgrammer from "C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer"
2. Run west flash, then I got below error
![image](https://user-images.githubusercontent.com/88422239/163803319-c3499cdc-9cb1-447d-8646-23d3137f5a09.png)
3. Install STM32CubeProgrammer at "C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer"
4.  Run west flash, then I got below error
![image](https://user-images.githubusercontent.com/88422239/163804642-1bfe8ca7-2301-411a-a1f6-1de037ee11b8.png)

STM32Cube Programmer version is V2.10.0
![image](https://user-images.githubusercontent.com/88422239/163804760-abda38c1-6706-47ce-aa84-aeda206e55cd.png)

> Also Be sure to have the ST link of the target board not connected elsewhere (by the windows CubeProgrammer application for example)

I have ensured that ST link of the target board not connected elsewhere. However I get the error even if target board is not connected.

**Comment by VineetaNarkhede-eaton** [comment, marked as answer]
**Problem is solved by installing older version STM32CubeProgrammer V2.7.0**
Fixed below issue by erasing device and ST Link firmware upgrade through STM32CubeProgrammer V2.7.0
![image](https://user-images.githubusercontent.com/88422239/163847607-b0f4677e-1126-4d47-ae55-53551eb600b6.png)

West flash:
![image](https://user-images.githubusercontent.com/88422239/163848061-30e75b2c-9817-4520-94cb-71f4b2573163.png)

Verified with zephyr version 3.0.0 and 3.0.99
